### PR TITLE
article added for testing-wai

### LINKED ISSUE #92: Article adding testing to a WAI package
Demonstrate how to use the `wasmer-pack-testing` crate, using the `sgp4` package as an example.

### PR DESCRIPTION
**Pull request by dynamite-bud**
Fixes #92.

### CODE REVIEW

**Comment by smiletondi** [comment]
image

Is it intended to be shown in that manner? @dynamite-bud

**Comment by dynamite-bud** [comment]
> image
> 
> Is it intended to be shown in that manner? @dynamite-bud

No @smiletondi , Fixed it bow.

File: _posts/testing-wai.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+date: '2023-03-14T11:25:55.785Z'
+author: Rudra
+published: true
+---
+
+# Testing - WAI
+
+![https://images.unsplash.com/photo-1597279393696-d5701aee7bf7?ixlib=rb-4.0.3&q=80&fm=jpg&crop=entropy&cs=tinysrgb](https://images.unsplash.com/photo-1597279393696-d5701aee7bf7?ixlib=rb-4.0.3&q=80&fm=jpg&crop=entropy&cs=tinysrgb)
+
+In my previous [article](https://wasmer.io/posts/WAI-is-the-answer), I explained how to port a package from Rust for the JavaScript ecosystem. These published packages need testing in their own ecosystems for the package verification and identification of any possible bugs.
```
Comment (ayys) [inline _posts/testing-wai.md:13]:
```suggestion
In my previous [article](https://wasmer.io/posts/WAI-is-the-answer), I explained how to port a Rust package for use in the JavaScript ecosystem. However, these published packages need to be tested within their own ecosystems to verify their functionality and identify any potential bugs.
```

File: _posts/testing-wai.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+## Testing in WAI
+
+Testing for WAI has been made easier using an automated testing framework `[wasmer-pack-testing](https://crates.io/crates/wasmer-pack-testing)`. This crate automatically discovers the tests for available language configurations for WAI and runs the tests correspondingly.
```
Comment (ayys) [inline _posts/testing-wai.md]:
```suggestion
Testing for WAI has become easier with the [wasmer-pack-testing](<https://crates.io/crates/wasmer-pack-testing>) automated testing framework. This crate automatically discovers tests for available language configurations for WAI and runs the tests accordingly.
```

File: _posts/testing-wai.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+It uses `jest` for JavaScript/Typescript and `pytest` for Python.
```
Comment (ayys) [inline _posts/testing-wai.md:19]:
```suggestion
wasmer-pack-testing uses `jest` for JavaScript/Typescript and `pytest` for Python.
```

File: _posts/testing-wai.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+This testing framework allows for both **logical checks** using unit tests and **code generation** using snapshot testing.
```
Comment (ayys) [inline _posts/testing-wai.md:21]:
```suggestion
This testing framework allows for both **logical checks** through unit tests and **code generation** using snapshot testing.
```

File: _posts/testing-wai.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+## Tutorial Outline
+
+- Introduction
+- Environment Setup
+- Bindings generation
+- Testing for JavaScript
+
+### Introduction
+
+In this tutorial, we will test the recently published `rustfft` package. This package demonstrates the portability and logical computation power of WebAssembly that can be harnessed using WAI.
```
Comment (ayys) [inline _posts/testing-wai.md:32]:
```suggestion
In this tutorial, we will test the recently published `rustfft` package, which demonstrates the portability and logical computation power of WebAssembly. WAI can be used to harness that power!
```

File: _posts/testing-wai.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+<aside style="background: #fef08a; border-radius: 0.4rem; padding: 4px; font-weight: 500;">
+<span style="background: white; padding: 4px; border-radius: 9999px; margin: 0 2px 0 0">💡</span> RustFFT is a high-performance FFT library written in pure Rust.
+</aside>
+
+### Environment Setup
+
+The default testing using `cargo test` only runs the tests that are written using the `test` macro in the library files. For running integration tests we can create a new file in a separate tests directory.
```
Comment (ayys) [inline _posts/testing-wai.md:38]:
```suggestion
The default testing using `cargo test` only runs the tests that are written using the `test` macro in the library files. To run integration tests, we can create a new file in a separate `tests` directory.
```

File: _posts/testing-wai.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+1. Create an integration tests file
+
+   ```bash
+   $ mkdir tests && touch tests/rustfft-integration-tests.rs
+   ```
+
+2. Let’s update our `cargo.toml` to tell it that we are providing a custom test harness.
```
Comment (ayys) [inline _posts/testing-wai.md:46]:
```suggestion
2. Let’s update our `cargo.toml` to indicate that we are providing a custom test harness.
```

File: _posts/testing-wai.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+   anyhow = "1.0.68"
+   tracing-subscriber = { version = "0.3.16", features = ["env-filter"] }
+   wasmer-pack-testing = "0.7.0"
+
+   [[test]]
+   name = "rustfft-integration-tests"
+   harness = false
+   ```
+
+3. Finally, set up the integration tests to generate bindings and test for the appropriate target.
```
Comment (ayys) [inline _posts/testing-wai.md:60]:
```suggestion
3. Finally, let's write an integration test to generate bindings and test them against the appropriate target.
```

File: _posts/testing-wai.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+           .with_env_filter(EnvFilter::from_default_env())
+           .init();
+       wasmer_pack_testing::autodiscover(env!("CARGO_MANIFEST_DIR"))?;
+       Ok(())
+   }
+   ```
+
+### Bindings Generation
+
+For our testing suite to discover which targets to generate bindings for let’s create a blank JavaScript test file in our crate directory (the one with `Cargo.toml`) that `autodiscover()` can use.
```
Comment (ayys) [inline _posts/testing-wai.md:78]:
```suggestion
To enable our testing suite to discover which targets to generate bindings for, let's create a blank JavaScript test file in the crate directory (the one with `Cargo.toml`) that `autodiscover()` can use.
```

File: _posts/testing-wai.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+Now let’s create a blank `rustfft.test.js` in our main directory.
```
Comment (ayys) [inline _posts/testing-wai.md:80]: By main directory here, do you mean the create directory from the other line?

File: _posts/testing-wai.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+```shell-session
+$ touch sgp4.test.js
+```
+
+Running the tests will generate a testing package (note the `package.json``) and bindings for your test to use.
```
Comment (ayys) [inline _posts/testing-wai.md:86]:
```suggestion
Running the tests will generate a testing package (note the `package.json`) and bindings for your test to use.
```

File: _posts/testing-wai.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+  ];
+  let dft = Algorithm.newDft(wasm, signal.length, 'forward');
+  let output = dft.compute(signal);
+  output.forEach((element, index) => {
+    expect(element.re.toPrecision(2)).toBe(spectrum[index].re.toPrecision(2));
+  });
+});
+```
+
+Running the `cargo test` now will automatically run the `jest` test for us.
```
Comment (ayys) [inline _posts/testing-wai.md:168]:
```suggestion
Running `cargo test` now will automatically run the `jest` test from above for us.
```

File: _posts/testing-wai.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+Test Suites: 1 passed, 1 total
+Tests:       2 passed, 2 total
+Snapshots:   0 total
+Time:        0.169 s
+Ran all test suites.
+✨  Done in 0.76s.
+
+```
+
+We see here that our tests are ran using the `jest` testing suite and pass for both the test cases.
```
Comment (ayys) [inline _posts/testing-wai.md:209]:
```suggestion
We see here that our tests run using the `jest` testing suite and and both test cases pass successfully.
```

File: _posts/testing-wai.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+## Conclusion
+
+In this tutorial, we got to know how to testing your published packages with WAI. The testing was done using JavaScript. This testing also concludes that our logic internally is working correctly and correct bindings are being generated for our package in JavaScript
```
Comment (ayys) [inline _posts/testing-wai.md:213]: I asked ChatGPT to generate a conclusion for this article. Looks pretty neat.



In conclusion, testing your published packages with WAI using JavaScript is an effective way to ensure that your package's internal logic is working correctly and that the correct bindings are being generated in JavaScript. With the [wasmer-pack-testing](<https://crates.io/crates/wasmer-pack-testing>) automated testing framework, testing for WAI has become easier than ever. This framework allows for both logical checks using unit tests and code generation using snapshot testing. By following the steps outlined in this tutorial, you can harness the portability and logical computation power of WebAssembly and ensure the quality of your published packages.

**Review by ayys — COMMENTED** [review]
@dynamite-bud This was a fun one to go through! I was planning on writing bindings for a rust project over the weekend, so this is very helpful to me. I added some minor changes which you can choose to incorporate (or not! These are just my suggestions).

File: _posts/testing-wai.md (hunk 3 above)
Comment (dynamite-bud) [reply _posts/testing-wai.md:19]: changed from it to wasmer-pack-testing

File: _posts/testing-wai.md (hunk 10 above)
Comment (dynamite-bud) [reply _posts/testing-wai.md:80]: yeah

**Comment by smiletondi** [comment]
image

Are the ticks also intended to be shown ? @dynamite-bud

**Comment by dynamite-bud** [comment]
@smiletondi yeah, the markdown renderer isn't turning the backticks to something like `this`

**Comment by smiletondi** [comment]
Okay, cool.

**Review by smiletondi — APPROVED** [review]

File: _posts/testing-wai.md [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+wasmer-pack-testing uses `jest` for JavaScript/Typescript and `pytest` for Python.
+
+This testing framework allows for both **logical checks** through unit tests and **code generation** using snapshot testing.
+
+## Tutorial Outline
+
+- Introduction
+- Environment Setup
+- Bindings generation
+- Testing for JavaScript
```
Comment (Michael-F-Bryan) [inline _posts/testing-wai.md:28]: Do we still need this list?

File: _posts/testing-wai.md [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+   ```
+
+### Bindings Generation
+
+To enable our testing suite to discover which targets to generate bindings for, let's create a blank JavaScript test file in the crate directory (the one with `Cargo.toml`) that `autodiscover()` can use.
+
+Now let’s create a blank `rustfft.test.js` in the crate directory.
+
+```shell-session
+$ touch sgp4.test.js
```
Comment (Michael-F-Bryan) [inline _posts/testing-wai.md:83]: Should they create `rustfft.test.js` or `sgp4.test.js`?

File: _posts/testing-wai.md (hunk 14 above)
Comment (ayys) [reply _posts/testing-wai.md]: Sure thing!

**Review by Michael-F-Bryan — APPROVED** [review]

**Review by smiletondi — APPROVED** [review]

**Review by smiletondi — APPROVED** [review]

**Pull request merged**